### ISSUE TITLE
How much preprocessing can we expect from the user / should we include into CPA?

### ISSUE THREAD

**Issue by daviddao** [CellProfiler/CellProfiler-Analyst#101]
Moved from #90 @AnneCarpenter : Do we need a deep discussion on whether feature reduction is necessary as a pre-processing step for some of the machine learning methods we now have? It wasn’t so useful for Gentle Boosting so it’s understandable this was not really a supported feature when that was (almost) the only option. But now that we have many other options - do those options include feature reduction within them, or do we need to offer this as pre-processing?

<em>Background</em>
Due to the fact, that we now provide new algorithms, many require preprocessed data. For example, SVMs only work well if the data is normalized and scaled.

<em>Discussion</em>
Can we expect the user to normalize and scale his data or should those algorithms do that automatically? How much data preprocessing can we expect, should we offer?

**Comment by LeeKamentsky** [comment]
This is what I would do:
* Dimensionality reduction creates a new database table with the reduced features and writes out a property file (with a new name) that replaces the old object table name with the dimensionality reduced and normalized one.
* The user is prompted to restart CPA to use the new property file.
* The user is encouraged to use classifiers appropriate for CP data via the interface and discouraged from using classifiers whose assumptions are not valid.

This keeps the code relatively simple by keeping the steps separate and by imposing a limited and localized burden on the code.

IMHO, some classifiers like Random Forest and Gentle Boosting account well for unbalanced distributions and redundant features and naive users should be guided towards those classifiers. SVM and others require so much domain knowledge that a user should have a well-thought-out reason for using them and a few extra steps and a transparent, but lengthy workflow should be an acceptable burden for experienced users and an appropriate barrier for naive ones.

**Comment by jhung0** [comment]
I think making a new properties file would be pretty cumbersome for the
user and hard to explain. Why not preprocess the training data
automatically without a button when training the classifier?

http://stackoverflow.com/questions/13324071/scaling-data-in-scikit-learn-svm
http://scikit-learn.org/stable/modules/preprocessing.html#standardization-or-mean-removal-and-variance-scaling

**Comment by daviddao** [comment]
I agree with @jhung0 that it is pretty cumbersome and hard to explain. But on the other hand, I don't like the idea to preprocess the data without telling the user. That would also give as less freedom in the way how to preprocess.

There are many ways how to preprocess the data. Just to name few: 
- Subtracting the mean or mode
- Scaling between 0 and 1 or -1 and 1
- Removing or keeping NaNs, 0s, low variance features. 

Each could have different reasoning and results depend highly on the given data and classifier.

Data preprocessing and feature engineering actually takes 80% of machine learning and there is no real rule of thumb.

**Comment by LeeKamentsky** [comment]
I just thought of something else. You have to package up the normalization and feature reduction steps in the pickle for CellProfiler. More and more complexity.

**Comment by shntnu** [comment]
@LeeKamentsky Regarding multiple steps, http://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html might come to one's rescue, like in https://github.com/CellProfiler/CellProfiler-Analyst/blob/master/cpa/profiling/decomp.py

**Comment by daviddao** [comment]
Our normalization tool now creates object tables with unnormalized X and Y location of the cells (8d90d36). Biologists can now normalize their object table (and in the same time do feature selection) and restart CPA with the normalized object table. Normalized data yield 2-3 % accuracy improvement for our CPA example dataset.

**Comment by AnneCarpenter** [comment]
It sounds like this is resolved @daviddao ? Re-open if not!
